AX: Implement new ARIA roles sectionfooter and sectionheader

rahimabdi opened the pull request:
#### 3d8ea36b3cc82c528c9a418846da2b6c8c976a53
<pre>
AX: Implement new ARIA roles sectionfooter and sectionheader
<a href="https://bugs.webkit.org/show_bug.cgi?id=273325">https://bugs.webkit.org/show_bug.cgi?id=273325</a>
<a href="https://rdar.apple.com/127119395">rdar://127119395</a>

Reviewed by Tyler Wilcock.

When a &lt;header&gt; or &lt;footer&gt; element with an accname is a descendant of a sectioning element (e.g., &lt;main&gt;, &lt;section&gt;), it should be
exposed as a named, navigable &quot;section header&quot; or &quot;section footer&quot; region respectively for assistive technology users. Both new, non-landmark roles
shouldn&apos;t be conflated with the existing document-level ARIA &quot;banner&quot; (&lt;header&gt; scoped to &lt;body&gt;) nor &quot;contentinfo&quot; (&lt;footer&gt; scoped to &lt;body&gt;) landmark roles.

This PR adds support for the new &quot;sectionfooter&quot; and &quot;sectionheader&quot; ARIA roles, and ensures that their computed role string (i.e., for interop testing) is
accurate: &lt;header&gt; returns &quot;sectionheader&quot; and &lt;footer&gt; returns &quot;sectionfooter&quot;.

* LayoutTests/accessibility/mac/footer-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/footer-roledescription-expected.txt:  Updated test expectations.
* LayoutTests/accessibility/mac/footer-roledescription.html: Updated test cases.
* LayoutTests/accessibility/mac/footer.html: Updated test cases.
* LayoutTests/accessibility/mac/header-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/header.html: Updated test cases.
* LayoutTests/accessibility/mac/html-section-elements-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/html-section-elements.html: Updated test cases.
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.tentative-expected.txt: Updated test expectations.
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/role/contextual-roles.tentative-expected.txt: Updated test expectations.
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/role/roles.tentative-expected.txt: Updated test expectations.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::ariaLandmarkRoleDescription const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::accessibilityRoleToString):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::accessibleNameDerivesFromContent const):
(WebCore::initializeRoleMap):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
* Source/WebCore/accessibility/atspi/AccessibilityAtspi.cpp:
* Source/WebCore/accessibility/atspi/AccessibilityAtspiEnums.h:
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::atspiRole):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformDescription):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(WebAccessibilityObjectWrapper determineIsAccessibilityElement)
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXARIAContentGroupText):

Canonical link: <a href="https://commits.webkit.org/296017@main">https://commits.webkit.org/296017@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd504ae0e46df8799124505a9920f8d91fd72bd6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17084 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112188 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/57540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27360 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35189 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81227 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/57540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109938 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21719 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61580 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21151 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14597 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56986 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91062 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14621 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115226 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34073 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25132 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90273 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34441 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92719 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89998 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22961 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34922 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12723 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29787 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33997 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39452 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33745 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37097 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35377 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -471,15 +471,23 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
+        // https://github.com/w3c/aria/pull/1931
+        // A <footer> that is a descendant of <main> or a sectioning element should have role="sectionfooter".
+        if (isDescendantOfElementType({ articleTag, asideTag, mainTag, navTag, sectionTag }))
+            return AccessibilityRole::SectionFooter;
```
Rather than adding a separate ancestry traversal, is it possible to combine this into one hand-rolled one, e.g. this:

```cpp
for (auto& ancestorElement : ancestorsOfType<Element>(*node)) {
    // check for tags here...
}
```

rahimabdi replied:
Definitely, I was unsure about this one because it's not a simple if/else construct (I'm checking a smaller subset of tags for the second traversal) so will streamline and combine into the existing `AccessibilityRole::Footer`.

twilco commented on Source/WebCore/accessibility/atspi/AccessibilityAtspiEnums.h:
```diff
@@ -135,6 +135,8 @@ enum class Role {
     Article,
     Landmark,
+    SectionFooter,
+    SectionHeader,
```
I know the existing alphabetization in this enum is a bit haphazard, but maybe we can try to put these two by the other "s" roles (maybe before "Static")

rahimabdi replied:
Sure thing (I thought putting it next to "Landmark" would be sensible since it's logically related).

rahimabdi committed:
AX: Implement new ARIA roles sectionfooter and sectionheader
https://bugs.webkit.org/show_bug.cgi?id=273325
rdar://127119395

Reviewed by Tyler Wilcock.

When a <header> or <footer> element with an accname is a descendant of a sectioning element (e.g., <main>, <section>), it should be
exposed as a named, navigable "section header" or "section footer" region respectively for assistive technology users. Both new, non-landmark roles
shouldn't be conflated with the existing document-level ARIA "banner" (<header> scoped to <body>) nor "contentinfo" (<footer> scoped to <body>) landmark roles.

This PR adds support for the new "sectionfooter" and "sectionheader" ARIA roles, and ensures that their computed role string (i.e., for interop testing) is
accurate: <header> returns "sectionheader" and <footer> returns "sectionfooter".

* LayoutTests/accessibility/mac/footer-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/footer-roledescription-expected.txt:  Updated test expectations.
* LayoutTests/accessibility/mac/footer-roledescription.html: Updated test cases.
* LayoutTests/accessibility/mac/footer.html: Updated test cases.
* LayoutTests/accessibility/mac/header-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/header.html: Updated test cases.
* LayoutTests/accessibility/mac/html-section-elements-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/html-section-elements.html: Updated test cases.
* LayoutTests/imported/w3c/web-platform-tests/html-aam/roles-contextual.tentative-expected.txt: Updated test expectations.
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/role/contextual-roles.tentative-expected.txt: Updated test expectations.
* LayoutTests/imported/w3c/web-platform-tests/wai-aria/role/roles.tentative-expected.txt: Updated test expectations.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::ariaLandmarkRoleDescription const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::accessibilityRoleToString):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::accessibleNameDerivesFromContent const):
(WebCore::initializeRoleMap):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
* Source/WebCore/accessibility/atspi/AccessibilityAtspi.cpp:
* Source/WebCore/accessibility/atspi/AccessibilityAtspiEnums.h:
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::atspiRole):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::rolePlatformDescription):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(WebAccessibilityObjectWrapper determineIsAccessibilityElement)
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXARIAContentGroupText):

Canonical link: https://commits.webkit.org/296017@main
